### PR TITLE
8330606: Redefinition doesn't but should verify the new klass

### PR DESCRIPTION
**Pull request by coleenp**
This change adds a couple of comments, removes some ancient bootstrapping code, and adds an old test that we call the verifier for redefining a class, even one in the bootclass path.

The fix for always verifying redefined classes was actually pushed with JDK-8341094, where the verifier code respected the parameter "should_verify_class".  By default this parameter follows the -Xverify setting.  For redefinition, this is passed as true.   The rest of the fix removes the special bootstrap loader cases that may have failed early on in the JDK development with -Xverify:all but now no longer do.  If someone tries to redefine these classes, they should also do the verification on the redefined bytecodes.

Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330606](https://bugs.openjdk.org/browse/JDK-8330606): Redefinition doesn't but should verify the new klass (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22116/head:pull/22116` \
`$ git checkout pull/22116`

Update a local copy of the PR: \
`$ git checkout pull/22116` \
`$ git pull https://git.openjdk.org/jdk.git pull/22116/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22116`

View PR using the GUI difftool: \
`$ git pr show -t 22116`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22116.diff">https://git.openjdk.org/jdk/pull/22116.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22116#issuecomment-2477228876)
</details>

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        return classWriter.toByteArray();
+    }
+
+    public static void main(String[] args) throws Exception {
+
+        Class<?> verifyErrorMirror = java.lang.VerifyError.class;
+
+        try {
+            RedefineClassHelper.redefineClass(verifyErrorMirror, dump());
+            VerifyError err = new VerifyError("verify me now");
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:124]: Which of these two lines are expected to throw a VerifyError?

Is the VerifyError because you can't redefine VerifyError, or is there something invalid about the how the class is redefined?

Comment (coleenp) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:124]: We call the verifier during redefinition with the 'true' parameter, so the redefinition will throw VerifyError when it's loading the new class file version.  There's a broken stackmap in the new version of this VerifyError class (ie. I asmifyied it and commented out something in it).  I updated the test with some comments.

Comment (plummercj) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java]: Ok. You got rid of the `new VerifyError()`, which is what I figured my next request was going to be. :)

The test looks fine. I don't really have the background to review the native code.

File: src/hotspot/share/classfile/verifier.cpp [hunk 2]
```diff
@@ -277,14 +279,6 @@ bool Verifier::is_eligible_for_verification(InstanceKlass* klass, bool should_ve
   Symbol* name = klass->name();
 
   return (should_verify_class &&
-    // return if the class is a bootstrapping class
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/verifier.cpp:280]: This method should have a comment explaining what the eligibility criteria is and how/why `should_verify_class` should be set in different conditions.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        ClassWriter classWriter = new ClassWriter(0);
+        FieldVisitor fieldVisitor;
+        RecordComponentVisitor recordComponentVisitor;
+        MethodVisitor methodVisitor;
+        AnnotationVisitor annotationVisitor0;
+
+        classWriter.visit(52, ACC_SUPER | ACC_PUBLIC, "java/lang/VerifyError", null, "java/lang/LinkageError", null);
+        {
+            fieldVisitor = classWriter.visitField(ACC_PRIVATE | ACC_FINAL | ACC_STATIC, "serialVersionUID", "J", null, new Long(7001962396098498785L));
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:68]: Why do we need a serialVersionUID in this test class?

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            fieldVisitor.visitEnd();
+        }
+        {
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:71]: Please add a comment showing the Java code this asm is implementing.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            methodVisitor.visitLineNumber(44, label1);
+            methodVisitor.visitInsn(RETURN);
+            Label label2 = new Label();
+            methodVisitor.visitLabel(label2);
+            methodVisitor.visitLocalVariable("this", "Ljava/lang/VerifyError;", null, label0, label2, 0);
+            methodVisitor.visitMaxs(1, 1);
+            methodVisitor.visitEnd();
+        }
+
+        {   // broken method
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:90]: Please add a comment showing the Java code this asm is implementing.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 4 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:71]: Actually why do you even need this constructor now?

**Review by dholmes-ora — COMMENTED** [review]
Sorry but I don't see how these changes deal with "redefinition should but doesn't verify the new klass" - the only change here is to not skip Object/Class/String/Throwable. ??

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 5 above)
Comment (coleenp) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:90]: This test came from asmifying java.lang.VerifyError and hacking it to have a VerifyError.  The code it's implementing is a lot of code and not interesting.  I can try cutting some things out.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 3 above)
Comment (coleenp) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:68]: TEST RESULT: Failed. Execution failed: `main' threw exception: java.lang.UnsupportedOperationException: class redefinition failed: attempted to change the schema (add/remove fields)

Because it's a field in VerifyError

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 4 above)
Comment (coleenp) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:72]: For redefinition, you can't add or delete methods.

File: src/hotspot/share/classfile/verifier.cpp [hunk 6]
```diff
@@ -105,11 +105,13 @@ static verify_byte_codes_fn_t verify_byte_codes_fn() {
 
 // Methods in Verifier
 
+// This method determines whether we run the verifier and class file format checking code.
 bool Verifier::should_verify_for(oop class_loader) {
   return class_loader == nullptr ?
     BytecodeVerificationLocal : BytecodeVerificationRemote;
 }
 
+// This method determines whether we allow package access in access checks in reflection.
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/classfile/verifier.cpp:114]: I added "This method" because the two methods right next to each other look the same but are different.  I want to say this one does one thing and this other one does something subtly different.  And it makes it a complete sentence.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 3 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:69]: Got it.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 4 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:72]: Got it.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 5 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java]: I hadn't quite grasped the basic operation of the test - I get it now. Thanks

File: src/hotspot/share/classfile/verifier.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/verifier.cpp:280]: I moved the function near should_be_verified, so it would be clear what this does and why, and added a comment.

**Comment by coleenp** [comment]
I added to the description to explain how the fix works.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    public static void main(String[] args) throws Exception {
+
+        Class<?> verifyErrorMirror = java.lang.VerifyError.class;
+
+        try {
+            // The Verifier is called for the redefinition, which will fail because of the broken <init> method above.
+            RedefineClassHelper.redefineClass(verifyErrorMirror, dump());
+            throw new RuntimeException("This should throw VerifyError");
+        } catch (VerifyError e) {
+            // JVMTI recreates the VerifyError so the verification message is lost.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:96]: I'm not clear why JVMTI would mess with the `VerifyError` that the verifier should have raised, and which explains the reason why verification failed. Not that your changes impact that.

File: src/hotspot/share/oops/method.cpp [hunk 8]
```diff
@@ -335,7 +335,8 @@ int Method::bci_from(address bcp) const {
 
 
 int Method::validate_bci(int bci) const {
-  return (bci == 0 || bci < code_size()) ? bci : -1;
+  // Called from the verifier, and should return -1 if not valid.
+  return ((is_native() && bci == 0) || (!is_native() && 0 <= bci && bci < code_size())) ? bci : -1;
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/method.cpp:339]: This expansion seems more correct, but it also seems unrelated to the main changes.

**Review by dholmes-ora — APPROVED** [review]
Thanks for explaining how the main fix was done elsewhere.

A couple of comments below but changes seem okay.

Thanks

File: src/hotspot/share/oops/method.cpp (hunk 8 above)
Comment (coleenp) [reply src/hotspot/share/oops/method.cpp:339]: This expansion was a bug that I hit while testing the main change.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 7 above)
Comment (coleenp) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:96]: The JVMTI code for redefinition returns JVMTI_ERROR_FAILS_VERIFICATION and not the pending exception to the caller.  The agent eventually in java.instrument/share/native/libinstrument/JavaExceptions.c will recreate the VerifyError based on the JVMTI error code.

File: src/hotspot/share/oops/method.cpp (hunk 8 above)
Comment (coleenp) [reply src/hotspot/share/oops/method.cpp:339]: I removed all the code for the VerifyError <init> code to reduce the test case, so the failing BCI was 0 and the verifier reporting tested whether this was a valid bytecode, then asserted a couple lines further that it wasn't a valid bytecode.

File: test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java (hunk 7 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:96]: Wow that is pretty developer unfriendly. Hopefully exception logging will still show what the error is in detail.

Comment (coleenp) [reply test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java:96]: Yes, logging does show the detailed verify error. The test has -Xlog:exceptions also that shows where it came from.

**Review by jdksjolen — APPROVED** [review]
LGTM

**Comment by coleenp** [comment]
Thanks David and Johan for the reviews.
/integrate

**Comment by theresa-m** [comment]
Hi @coleenp for the scenario written in `serviceability/jvmti/RedefineClasses/RedefineVerifyError.java` I believe the correct behavior should be throwing a ClassFormatException since the generated class is not well formed.

According to the JVM specification in section `4.7.3 The Code Attribute`:

```
If the method is either native or abstract, and is not a class or interface
initialization method, then its method_info structure must not have a Code attribute
in its attributes table. Otherwise, its method_info structure must have exactly
one Code attribute in its attributes table.
```
also
`The value of code_length must be greater than zero`

Can you clarify why VerifyError is expected to be caught here?

**Comment by offamitkumar** [comment]
Hi @coleenp, 

I am repeating the message from Theresa, since she is part of OpenJ9 and can't accept  OpenJDK Terms: 

```
`serviceability/jvmti/RedefineClasses/RedefineVerifyError.java` I believe the correct behavior should be throwing a ClassFormatException since the generated class is not well formed.

According to the JVM specification in section 4.7.3 The Code Attribute:

If the method is either native or abstract, and is not a class or interface
initialization method, then its method_info structure must not have a Code attribute
in its attributes table. Otherwise, its method_info structure must have exactly
one Code attribute in its attributes table.
also
The value of code_length must be greater than zero

Can you clarify why VerifyError is expected to be caught here?
```

**Comment by dholmes-ora** [comment]
Hi Amit (@offamitkumar ),

I think the answer to that is already given above:

https://github.com/openjdk/jdk/pull/22116#discussion_r1850265592

> The JVMTI code for redefinition returns JVMTI_ERROR_FAILS_VERIFICATION and not the pending exception to the caller. The agent eventually in java.instrument/share/native/libinstrument/JavaExceptions.c will recreate the VerifyError based on the JVMTI error code.

**Comment by coleenp** [comment]
Thanks for answering David.  The new class should get a VerifyError not a ClassFormatError - the bytecodes were incorrect, as detected by the verifier.  It's not missing a code attribute but the code is truncated as reported with -Xlog:exceptions:

```
[0.339s][info][exceptions] Exception <a 'java/lang/VerifyError'{0x00000007ff8850c8}: Control flow falls through code end
[                        ] Exception Details:
[                        ]   Location:
[                        ]     java/lang/VerifyError.<init>()V @0: <invalid>
[                        ]   Reason:
[                        ]     Error exists in the bytecode
[                        ]   Bytecode:
[                        ]     0000000:
[                        ] > (0x00000007ff8850c8)
[                        ] thrown [/scratch/cphillim/hg/jdk-ci-pd/open/src/hotspot/share/classfile/verifier.cpp, line 293]
[                        ] for thread 0x000075ea2cbbe260
```

In the test the methods are here just don't have anything in them:

```
        {
            methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
            methodVisitor.visitCode();
            methodVisitor.visitEnd();
        }

        {
            methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "(Ljava/lang/String;)V", null, null);
            methodVisitor.visitCode();
            classWriter.visitEnd();
        }
```

So VerifyError is the right exception.  It might be that you need to add some nop() bytecode or something to generate the code attribute?

**Comment by offamitkumar** [comment]
>  It might be that you need to add some nop() bytecode or something to generate the code attribute?

@coleenp This seems like a valid solution. By the way, can we modify this testcase in OpenJDK or should the OpenJ9 team make the change in their repository. Which one do you suggest ?
